### PR TITLE
Supporter+ 2023 Migration V1 to V2 

### PR DESCRIPTION
**Pull request by shtukas**
Implementation of the supporter plus migration from version 1 to version 2, essentially a change of rate plans, without price increase. 

This migration is different from the previous ones because there indeed in no price increase (with that said there is still emails to be sent to inform the subscribers of a change in terms and conditions) and also because the target rate plan has two charges instead of one.

Considering the slightly more complex logic, we have here thoroughly tested 4 subscriptions

1. Monthly Standard: a monthly subscription with a charge of 10 GBP, meaning doesn't have an extra contribution charge. The two important things to verify are how it evaluates and the amendment data. 

We evaluate as

```
PriceData(currency, 10, 10, "Month")

SuccessfulEstimationResult(
  subscriptionName = "SUBSCRIPTION-NUMBER",
  startDate = LocalDate.of(2023, 9, 1),
  currency = "GBP",
  oldPrice = BigDecimal(10.0),
  estimatedNewPrice = BigDecimal(10),
  billingPeriod = "Month"
)
```

which shows that prices before and after equal the rate plan(s). Note that in any other migration, the subscription would be given the processing stage `NoPriceIncrease`, but we modified that rule just for this migration. 

The update is then

```
ZuoraSubscriptionUpdate(
   add = List(AddZuoraRatePlan("8a128ed885fc6ded01860228f77e3d5a", LocalDate.of(2023, 8, 1))),
   remove = List(RemoveZuoraRatePlan("8a129d388962fe000189679374ce42a1", LocalDate.of(2023, 8, 1))),
   currentTerm = None,
   currentTermPeriodType = None
)
```

2. Monthly with Contribution. Here we have a 10 GBP rate plan with a 15 USD contribution totalling to 25 USD. 

We evaluate at 

```
PriceData(currency, 25, 10, "Month")

SuccessfulEstimationResult(
  subscriptionName = "SUBSCRIPTION-NUMBER",
  startDate = LocalDate.of(2023, 9, 3),
  currency = "GBP",
  oldPrice = BigDecimal(25.0),
  estimatedNewPrice = BigDecimal(10),
  billingPeriod = "Month"
)
```

The "old" price is correctly computed as 25 GBP (which is all that the customer is paying) , but note here that the new price is 10 GBP and is the charge of the new rate plan itself without contribution. 

The update is then

```
ZuoraSubscriptionUpdate(
  add = List(
    AddZuoraRatePlan(
      "8a128ed885fc6ded018602296ace3eb8",
      LocalDate.of(2023, 8, 3),
      chargeOverrides = List(
        ChargeOverride(
          productRatePlanChargeId = "8a128d7085fc6dec01860234cd075270",
          billingPeriod = "Month",
          price = 15
        )
      )
    )
  ),
  remove = List(RemoveZuoraRatePlan("8a12921d89018aaa01891bef52021b65", LocalDate.of(2023, 8, 3))),
  currentTerm = None,
  currentTermPeriodType = None
)
```

Here we add the intended target rate plan and we also define a 15 (GBP) charge override for the contribution rate plan.

3. Annual standard: An annual subscription with standard rate plan charge 120 USD. It evaluates as 

```
PriceData(currency, 120, 120, "Annual")

SuccessfulEstimationResult(
  subscriptionName = "SUBSCRIPTION-NUMBER",
  startDate = LocalDate.of(2024, 7, 2),
  currency = "USD",
  oldPrice = BigDecimal(120),
  estimatedNewPrice = BigDecimal(120),
  billingPeriod = "Annual"
)
```

The update is then

```
ZuoraSubscriptionUpdate(
  add = List(AddZuoraRatePlan("8a128ed885fc6ded01860228f77e3d5a", LocalDate.of(2024, 7, 2))),
  remove = List(RemoveZuoraRatePlan("8a128432890171d1018914866bee0e7f", LocalDate.of(2024, 7, 2))),
  currentTerm = None,
  currentTermPeriodType = None
)
```

4. And last but not least, an annual subscription paying more than the related v1 rate plan. The v1 rate plan is at 95 GBP, and the subscription pays 120 GBP. It evaluates at 

```
PriceData(currency, 120, 95, "Annual")

SuccessfulEstimationResult(
  subscriptionName = "SUBSCRIPTION-NUMBER",
  startDate = LocalDate.of(2024, 6, 28),
  currency = "GBP",
  oldPrice = BigDecimal(120),
  estimatedNewPrice = BigDecimal(95),
  billingPeriod = "Annual"
)
```

Here again the old price is correctly computed as 120 GBP, but the espimated new price is that of the target rate plan of 95 GBP without the extra 25 contribution. 

The update is then

```
ZuoraSubscriptionUpdate(
  add = List(
    AddZuoraRatePlan(
      "8a128ed885fc6ded01860228f77e3d5a",
      LocalDate.of(2024, 6, 28),
      chargeOverrides = List(
        ChargeOverride(
          productRatePlanChargeId = "8a12892d85fc6df4018602451322287f",
          billingPeriod = "Annual",
          price = 25
        )
      )
    )
  ),
  remove = List(RemoveZuoraRatePlan("8a12843288f6ded10188ff5fbef67bb3", LocalDate.of(2024, 6, 28))),
  currentTerm = None,
  currentTermPeriodType = None
)
```

As for the case of the monthly subscription with extra contribution, we target the v2 rate plan and provision a 25 GBP contribution override.

### CODE REVIEW

File: lambda/src/test/resources/SupporterPlus2023V1V2/monthly-standard/account.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    "accountNumber" : "ACCOUNT-NUMBER",
+    "notes" : null,
+    "status" : "Active",
+    "crmId" : "0010J00001w3XTWQA2",
+    "batch" : "Batch1",
+    "invoiceTemplateId" : "2c92a0fb49377eae014951ca848e074b",
+    "communicationProfileId" : null,
+    "purchaseOrderNumber" : null,
+    "customerServiceRepName" : null,
+    "IdentityId__c" : "100560834",
```
Comment (graham228221) [inline lambda/src/test/resources/SupporterPlus2023V1V2/monthly-standard/account.json:14]: Another Identity Id

File: lambda/src/main/scala/pricemigrationengine/handlers/NotificationHandler.scala [hunk 2]
```diff
@@ -162,9 +166,9 @@ object NotificationHandler extends CohortHandler {
       forceEstimated = MigrationType(cohortSpec) match {
         case Membership2023Monthlies => true
         case Membership2023Annuals   => true
+        case SupporterPlus2023V1V2   => true
         case _                       => false
```
Comment (rupertbates) [inline lambda/src/main/scala/pricemigrationengine/handlers/NotificationHandler.scala:170]:
```suggestion
        case Legacy                       => false
```

File: lambda/src/main/scala/pricemigrationengine/handlers/SalesforcePriceRiseCreationHandler.scala [hunk 3]
```diff
@@ -84,6 +84,7 @@ object SalesforcePriceRiseCreationHandler extends CohortHandler {
       val forceEstimated = MigrationType(cohortSpec) match {
         case Membership2023Monthlies => true
         case Membership2023Annuals   => true
+        case SupporterPlus2023V1V2   => true
         case _                       => false
```
Comment (rupertbates) [inline lambda/src/main/scala/pricemigrationengine/handlers/SalesforcePriceRiseCreationHandler.scala:88]:
```suggestion
        case Legacy                       => false
```

File: lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+  }
+
+  def currencyToNewPriceAnnuals(currency: String): Either[AmendmentDataFailure, BigDecimal] = {
+    newPriceMapAnnuals.get(currency) match {
+      case None => Left(AmendmentDataFailure(s"Could not determine a new annual price for currency: ${currency}"))
+      case Some(price) => Right(price)
+    }
+  }
+
+  def currencyToNewPrice(bilingP: String, currency: String): Either[AmendmentDataFailure, BigDecimal] = {
```
Comment (rupertbates) [inline lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala:88]: typo, should be `billingP`

File: lambda/src/test/resources/SupporterPlus2023V1V2/annual-standard/account.json [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    "accountNumber" : "ACCOUNT-NUMBER",
+    "notes" : null,
+    "status" : "Active",
+    "crmId" : "0010J00001gz4tmQAA",
+    "batch" : "Batch1",
+    "invoiceTemplateId" : "2c92a0fb49377eae014951ca848e074b",
+    "communicationProfileId" : null,
+    "purchaseOrderNumber" : null,
+    "customerServiceRepName" : null,
+    "IdentityId__c" : "17583003",
```
[the remark this answers is not in the document]
Comment (shtukas) [reply lambda/src/test/resources/SupporterPlus2023V1V2/annual-standard/account.json]: Oh yeah. Thanks. I usually clean them up, but forgot that one :)

File: lambda/src/test/resources/SupporterPlus2023V1V2/monthly-standard/account.json (hunk 1 above)
Comment (shtukas) [reply lambda/src/test/resources/SupporterPlus2023V1V2/monthly-standard/account.json]: idem

File: lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        Left(
+          AmendmentDataFailure(
+            s"Subscription ${subscription.subscriptionNumber} has no rate plan charges"
+          )
+        )
+      }
+      case _ => {
+        Right(ratePlanCharges.map(x => x.price).filter(p => p.isDefined).map(p => p.get).fold(BigDecimal(0)) { (z, i) =>
+          z + i
+        })
```
Comment (rupertbates) [inline lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala]:
```suggestion
        Right(ratePlanCharges.flatMap(_.price).sum)
```

Comment (shtukas) [reply lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala]: Thanks. Corrected!

File: lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala (hunk 4 above)
Comment (shtukas) [reply lambda/src/main/scala/pricemigrationengine/model/SupporterRevenue2023V1V2.scala]: Corrected. Thanks!

**Comment by shtukas** [comment]
@graham228221 @rupertbates all done. Thanks!

**Review by rupertbates — APPROVED** [review]
Nice work 👍

**Comment by shtukas** [comment]
Corrected a bug: https://github.com/guardian/price-migration-engine/pull/958

**Pull request merged**